rfcs/0000-windows-containers.md

```diff
@@ -0,0 +1,136 @@
+
+If a user creates a cluster stack with build and run images of mixed os, the resource should be marked as not-ready
+
+**Actions to Take**
+
+- Add these validations to the kpack-controller
+- Add the label `os.kpack.io: windows` to each resource for os disambiguation
+
+### kpack should support creating windows builders
+
+This is currently supported in the [pack cli](https://github.com/buildpacks/pack/issues/469)
+
+**Actions to Take**
+
+- kpack-controller will use a new environment variable `WINDOWS_LIFECYCLE_IMAGE` to reference the windows lifecycle image
+- When a Builder or ClusterBuilder is created using a windows stack and windows buildpackages, kpack-controller must create a windows builder using the windows lifecycle image.
+    - Image os is determined by the image [os property](https://github.com/opencontainers/image-spec/blob/master/config.md#:~:text=os%20string).
+
+Notes:
+
+- Builders with mixed os images must error gracefully
+- Builders should be labeled with `os.kpack.io: windows` for os disambiguation independent of image metadata
+
+### kpack should support running builds on windows workers
+
+**Actions to Take**
+
+- kpack-controller will accept new environment variables for windows images needed to run builds. These will have the same functionality as on linux and will be used for running builds on windows workers.
+    - `WINDOWS_BUILD_INIT_IMAGE` - a windows build init image
+    - `WINDOWS_COMPLETION_IMAGE` - a windows completion image
+    - `WINDOWS_REBASE_IMAGE` - a windows rebase image
+    - These will be optional to remove the dependency on windows images
+- When an Image is created using a windows Builer or ClusterBuilder labeled with `os.kpack.io: windows`, it will create a build with new properties
+    - The Build will use `WINDOWS_BUILD_INIT_IMAGE` and `WINDOWS_COMPLETION_IMAGE` for the respective containers.
+    - The Build will be labeled with `os.kpack.io: windows` for os disambiguation
+ - When a Build is created with the label `os.kpack.io: windows`, it will use the node selector
+
+ ```
+ nodeSelector:
+     kubernetes.io/os: windows
+ ```
+
+## Complexity:
+
+High
+
+## Prior Art:
+
+[pack cli](https://github.com/buildpacks/pack) has support for windows containers
+
+## Risks:
+
+**Windows images can only run and be created on workers with the same OS version**
+
+Windows images can only run on windows nodes with the same OS. ex: Windows Server LTSC 2019 can only run images that are ltsc2019 or equivalent version [docs](https://docs.microsoft.com/en-us/virtualization/windowscontainers/deploy-containers/version-compatibility?tabs=windows-server-2019%2Cwindows-10-20H2#:~:text=Supports%20process%20isolation)
+
+Ergo:
+
+- Each windows image installed with `kpack` (`build-init`, `rebase`, `completion`, `lifecycle`) must be for the same os version and must match the cluster's windows nodes
+- Stacks and buildpackages must be for the same os version and must match the cluster's windows nodes
```

Comment:
All pack-generated Windows buildpackage images have a blank `os.version`, as currently implemented. Windows daemons give special treatment to a blank `os.version` - all known Windows daemon's will skip validating against their own version and are able to pull them. So `kpack` won't need to validate buildpackage's `os.version`. The only buildpackage-to-stack compatibility restrictions would be expressed as mixins, just like Linux, and should *just work*.

Windows Stack images (build/run) **do** always have `os.version` field (ex: `10.0.17763.1518` format: [`<major>.<minor>.<build>.<revision>`](https://docs.microsoft.com/en-us/dotnet/api/system.version.build?view=netcore-3.1)), set by Microsoft on the base image, that must match the os-version of the daemon where they are run.

Reply:
nice!